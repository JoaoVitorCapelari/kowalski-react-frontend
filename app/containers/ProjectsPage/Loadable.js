/**
 *
 * Asynchronously loads the component for ProjectsPage
 *
 */

import Loadable from 'react-loadable';

export default Loadable({
  loader: () => import('./index'),
  loading: () => null,
});
