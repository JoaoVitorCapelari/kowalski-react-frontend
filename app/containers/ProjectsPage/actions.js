/*
 *
 * ProjectsPage actions
 *
 */

import {
  DEFAULT_ACTION,
  OPEN_NEW_PROJECT_FORM,
  CLOSE_NEW_PROJECT_FORM,
  SUBMIT_NEW_PROJECT_FORM,
  NEW_PROJECT_SAVED,
  SUBMIT_NEW_PROJECT_FORM_AND_CLOSE_IT,
} from './constants';

export function defaultAction() {
  return {
    type: DEFAULT_ACTION,
  };
}

export function openNewProjectForm() {
  return {
    type: OPEN_NEW_PROJECT_FORM,
  };
}

export function closeNewProjectForm() {
  return {
    type: CLOSE_NEW_PROJECT_FORM,
  };
}

export function submitNewProjectForm(formData) {
  return {
    type: SUBMIT_NEW_PROJECT_FORM,
    payload: formData,
  };
}

export function submitNewProjectFormAndCloseIt(formData) {
  return {
    type: SUBMIT_NEW_PROJECT_FORM_AND_CLOSE_IT,
    payload: formData,
  };
}

export function newProjectSaved(newProjectData) {
  return {
    type: NEW_PROJECT_SAVED,
    payload: newProjectData,
  };
}
