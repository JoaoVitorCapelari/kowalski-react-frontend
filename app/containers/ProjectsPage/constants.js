/*
 *
 * ProjectsPage constants
 *
 */

export const DEFAULT_ACTION = 'app/ProjectsPage/DEFAULT_ACTION';
export const OPEN_NEW_PROJECT_FORM = 'app/ProjectsPage/OPEN_NEW_PROJECT_FORM';
export const CLOSE_NEW_PROJECT_FORM = 'app/ProjectsPage/CLOSE_NEW_PROJECT_FORM';
export const SUBMIT_NEW_PROJECT_FORM = 'app/ProjectsPage/SUBMIT_NEW_PROJECT_FORM';
export const SUBMIT_NEW_PROJECT_FORM_AND_CLOSE_IT = 'app/ProjectsPage/SUBMIT_NEW_PROJECT_FORM_AND_CLOSE_IT';
export const NEW_PROJECT_SAVED = 'app/ProjectsPage/NEW_PROJECT_SAVED';
export const NEW_PROJECT_FORM_ID = 'newProjectForm';
