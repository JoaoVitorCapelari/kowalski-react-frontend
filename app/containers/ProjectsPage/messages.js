/*
 * ProjectsPage Messages
 *
 * This contains all the text for the ProjectsPage component.
 */
import { defineMessages } from 'react-intl';

export default defineMessages({
  header: {
    id: 'app.containers.ProjectsPage.header',
    defaultMessage: 'This is ProjectsPage container !',
  },
});
