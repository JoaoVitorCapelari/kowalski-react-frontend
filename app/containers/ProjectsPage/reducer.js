/*
 *
 * ProjectsPage reducer
 *
 */

import { fromJS, List } from 'immutable';
import {
  OPEN_NEW_PROJECT_FORM,
  CLOSE_NEW_PROJECT_FORM,
  START_PROJECTS_LOADING,
  END_PROJECTS_LOADING,
} from './constants';

const initialState = fromJS({
  projects: [],
  projectFormOpen: false,
  isLoadingProjects: false,
  loadingProjectsErrorMsg: '',
});

function projectsPageReducer(state = initialState, { type, payload }) {
  switch (type) {
    case OPEN_NEW_PROJECT_FORM:
      return state.set('projectFormOpen', true);
    case CLOSE_NEW_PROJECT_FORM:
      return state.set('projectFormOpen', false);
    case START_PROJECTS_LOADING:
      return state.set('isLoadingProjects', true);
    case END_PROJECTS_LOADING:
      return state
        .set('isLoadingProjects', false)
        .set('loadingProjectsErrorMsg', payload.success ? '' : payload.errorMsg)
        .set('projects', payload.success ? payload.data : state.get('projects'));
    default:
      return state;
  }
}

export default projectsPageReducer;
