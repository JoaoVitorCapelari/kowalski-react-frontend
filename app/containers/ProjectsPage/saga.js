import { takeEvery, call, put } from 'redux-saga/effects';
import { push } from 'react-router-redux';
import {
  reset,
  stopSubmit,
  startSubmit,
} from 'redux-form';
import { fromJS } from 'immutable';
import {
  startProjectLoading,
  closeNewProjectForm,
  endProjectLoading,
} from './actions';

import {
  expiredSessionDetected,
} from '../App/actions';

import {
  SUBMIT_NEW_PROJECT_FORM,
  SUBMIT_NEW_PROJECT_FORM_AND_CLOSE_IT,
  NEW_PROJECT_FORM_ID,
  CLOSE_NEW_PROJECT_FORM,
  PROJECT_SELECTED,
  START_PROJECTS_LOADING,
} from './constants';
import { getProjects, createProject } from '../../support/backend/KowalskiBackendClient';
import { SERVER_BASE_URL } from '../../utils/constants';

export function* submitProjectForm({ payload }) {
  yield put(startSubmit(NEW_PROJECT_FORM_ID));
  try {
    yield call(createProject, {
      config: { baseUrl: SERVER_BASE_URL },
      token: localStorage.getItem('authToken'),
      projectData: payload.toJSON(),
    });
    yield put(stopSubmit(NEW_PROJECT_FORM_ID));
    yield put(startProjectLoading());
    return true;
  } catch (e) {
    yield put(stopSubmit(
      NEW_PROJECT_FORM_ID,
      { _error: 'Houve um erro enquanto se tentava comunicar com o servidor =(' },
    ));
  }

  return false;
}

export function* handleSubmit({ payload }) {
  const successful = yield submitProjectForm({ payload });
  if (successful) {
    yield clearProjectForm();
  }
}

export function* handleSubmitAndCloseForm({ payload }) {
  const successful = yield submitProjectForm({ payload });
  if (successful) {
    yield put(closeNewProjectForm());
  }
}

export function* clearProjectForm() {
  yield put(reset(NEW_PROJECT_FORM_ID));
}

export function* handleProjectSelected({ payload }) {
  yield put(push(`/projects/${payload}`));
}

export function* handleProjectLoading() {
  try {
    const projects = yield call(
      getProjects,
      { config: { baseUrl: SERVER_BASE_URL }, token: localStorage.getItem('authToken') },
    );
    yield put(endProjectLoading({ success: true, data: fromJS(projects) }));
  } catch (e) {
    if (e.response && e.response.status) {
      if (e.response.status === 403) {
        yield put(endProjectLoading({ success: false, errorMsg: null }));
        yield put(expiredSessionDetected());
        return;
      }
    }
    yield put(endProjectLoading({ success: false, errorMsg: 'Houve um erro enquanto se tentava comunicar com o servidor =(' }));
  }
}

export default function* defaultSaga() {
  yield takeEvery(SUBMIT_NEW_PROJECT_FORM, handleSubmit);
  yield takeEvery(SUBMIT_NEW_PROJECT_FORM_AND_CLOSE_IT, handleSubmitAndCloseForm);
  yield takeEvery(CLOSE_NEW_PROJECT_FORM, clearProjectForm);
  yield takeEvery(PROJECT_SELECTED, handleProjectSelected);
  yield takeEvery(START_PROJECTS_LOADING, handleProjectLoading);
}
