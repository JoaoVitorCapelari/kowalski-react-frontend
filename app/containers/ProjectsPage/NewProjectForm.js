import React from 'react';
import styled from 'styled-components';
import PropType from 'prop-types';
import { Field, reduxForm } from 'redux-form/immutable';
import { injectIntl, FormattedMessage } from 'react-intl';
import InputField from 'components/InputField/Loadable';
import TextAreaField from 'components/TextAreaField/Loadable';

import { required } from '../../support/forms/validation';
import { NEW_PROJECT_FORM_ID } from './constants';
import { formatDate } from '../../support/backend/formatters';
import messages from './messages';

const Wrapper = styled.div`
  display: flex;
  flex-grow: 1;
  flex-direction: column;
  justify-content: center;
  align-content: center;
`;

const FormTitle = styled.div`
`;

const FormActions = styled.div`
  display: flex;
  justify-content: center;
  margin: 2rem;
`;

const H1 = styled.h1`
  text-align: center;
`;

const FormAction = styled.button`
  font-size: 1.1rem;
  width: 100%;
`;

const FormActionWrapper = styled.div`
  width: 100%;
`;

function NewProjectForm(props) {
  const { error, submitting, onCancel, handleSubmit, onAdd, onSaveAndAddNew, intl: { formatMessage } } = props;

  const myOnSubmit = (formData, func) => (
    func(formData
      .set('startDate', formatDate(new Date()))
      .set('endDate', formatDate(new Date()))
    )
  );

  return (
    <Wrapper className="box">
      <form onSubmit={handleSubmit((formData) => myOnSubmit(formData, onAdd))}>
        <FormTitle><H1>{formatMessage(messages.newProjectFormTitle)}</H1></FormTitle>

        <Field
          name="code"
          id="code"
          component={InputField}
          label={formatMessage(messages.newProjectLabelCode)}
          validate={[required]}
          fieldStyle={{ width: '200px' }}
        />

        <Field
          name="name"
          id="name"
          component={InputField}
          label={formatMessage(messages.newProjectLabelName)}
          validate={[required]}
        />

        <Field
          name="description"
          id="description"
          component={TextAreaField}
          label={formatMessage(messages.newProjectLabelDescription)}
          validate={[required]}
        />

        {
          error &&
          <div className="control" style={{ marginTop: '1rem' }}>
            <article className="message is-danger">
              <div className="message-body">
                { error }
              </div>
            </article>
          </div>
        }

        <FormActions>
          <FormActionWrapper className="control">
            <FormAction
              className="button"
              type="button"
              disabled={submitting}
              onClick={handleSubmit((formData) => myOnSubmit(formData, onSaveAndAddNew))}
            >
              <FormattedMessage {... messages.newProjectButtonSaveAndAddNew} />
            </FormAction>
          </FormActionWrapper>
          <FormActionWrapper className="control">
            <FormAction type="submit" className={`button is-primary ${submitting ? 'is-loading' : ''}`} disabled={submitting}>
              <FormattedMessage {... messages.newProjectButtonAdd} />
            </FormAction>
          </FormActionWrapper>
          <FormActionWrapper className="control">
            <FormAction className="button" type="button" onClick={onCancel}>
              <FormattedMessage {... messages.newProjectButtonCancel} />
            </FormAction>
          </FormActionWrapper>
        </FormActions>
      </form>
    </Wrapper>
  );
}

NewProjectForm.propTypes = {
  error: PropType.any,
  submitting: PropType.any.isRequired,
  onCancel: PropType.func,
  handleSubmit: PropType.func,
  onAdd: PropType.func,
  onSaveAndAddNew: PropType.func,
  intl: PropType.object.isRequired,
};

export default reduxForm({
  form: NEW_PROJECT_FORM_ID,
})(injectIntl(NewProjectForm));
