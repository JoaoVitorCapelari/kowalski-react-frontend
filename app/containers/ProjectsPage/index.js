/**
 *
 * ProjectsPage
 *
 */

import React from 'react';
import PropTypes from 'prop-types';
import styled from 'styled-components';
import { connect } from 'react-redux';
import { Helmet } from 'react-helmet';
// import { FormattedMessage } from 'react-intl';
import { createStructuredSelector } from 'reselect';
import { List } from 'immutable';
import { compose } from 'redux';

import injectSaga from 'utils/injectSaga';
import injectReducer from 'utils/injectReducer';
import Modal from 'components/Modal/Loadable';
import NewProjectForm from './NewProjectForm';
import {
  makeSelectProjects,
  makeSelectIsNewProjectFormOpen,
  makeSelectIsSubmittingNewProject,
  makeSelectProjectCodes,
} from './selectors';
import reducer from './reducer';
import saga from './saga';
import * as actions from './actions';
// import messages from './messages';

const MainContainerWrapper = styled.div`
  display: flex;
  flex-direction: column;
`;

const renderProjects = (projects, onSelectProject) => {
  return (projects || []).map((project, index) => {
    // TODO trocar por
    return (
      <a
        tabIndex={index}
        className="tile is-child projectBox"
        role="button"
        key={project.get('name')}
        style={{ display: 'block' }}
        onClick={() => { onSelectProject(project.get('code')); }}
      >
        <p className="title">{project.get('name')}</p>
        <p>{project.get('description')}</p>
      </a>
    );
  });
};

const margins = `
  margin-top: 2rem;
  margin-left: 4rem;
  margin-right: 4rem;
`;

const TitleBar = styled.div`
  ${margins}
  display: flex;
  flex-basis: 100%;
  align-items: center;
  justify-content: space-between;
`;

const PageTitle = styled.h1`
  display: block;
  font-weight:300px;
  margin-left:3%;
`;

const AddProjectButton = styled.button`
width:226px;
height:50px !important;
border-radius:2px;
border:1px solid #654EA3;
color: #654EA3;
`;

const ProjectListWrapper = styled.div`
  ${margins}
  display: flex;
  flex-grow: 1;
  flex-direction: column;
  
`;

function ProjectsPage(props) {
  const {
    projects,
    projectSelected,
    openNewProjectForm,
    closeNewProjectForm,
    isNewProjectFormOpen,
  } = props;

  return (
    <MainContainerWrapper className="kowalski-react-basic-container">
      <Helmet>
        <title>Projetos</title>
        <meta name="description" content="Description of ProjectsPage" />
      </Helmet>
      <TitleBar>
        <PageTitle>Projects</PageTitle>
        <AddProjectButton className="button" onClick={openNewProjectForm}>Add project</AddProjectButton>
      </TitleBar>
      <ProjectListWrapper>
        <div className="tile is-parent is-vertical">
          { renderProjects(projects, projectSelected) }
        </div>
      </ProjectListWrapper>
      <Modal active={isNewProjectFormOpen} onDismiss={closeNewProjectForm}>
        <NewProjectForm
          onAdd={props.submitNewProjectFormAndCloseIt}
          onSaveAndAddNew={props.submitNewProjectForm}
          isSubmitting={props.isSubmittingNewProject}
          onCancel={closeNewProjectForm}
        />
      </Modal>
    </MainContainerWrapper>
  );
}

ProjectsPage.propTypes = {
  projects: PropTypes.instanceOf(List),
  openNewProjectForm: PropTypes.func,
  isNewProjectFormOpen: PropTypes.bool,
  closeNewProjectForm: PropTypes.func,
  submitNewProjectForm: PropTypes.func,
  isSubmittingNewProject: PropTypes.bool,
  submitNewProjectFormAndCloseIt: PropTypes.func,
  projectSelected: PropTypes.func,
};

const mapStateToProps = createStructuredSelector({
  projects: makeSelectProjects(),
  isNewProjectFormOpen: makeSelectIsNewProjectFormOpen(),
  isSubmittingNewProject: makeSelectIsSubmittingNewProject(),
});

const withConnect = connect(mapStateToProps, actions);

const withReducer = injectReducer({ key: 'projectspage', reducer });
const withSaga = injectSaga({ key: 'projectspage', saga });

export default compose(
  withReducer,
  withSaga,
  withConnect,
)(ProjectsPage);
