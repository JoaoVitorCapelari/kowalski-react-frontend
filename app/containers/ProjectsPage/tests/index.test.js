import React from 'react';
import { Provider } from 'react-redux';
import { mount } from 'enzyme';
import { IntlProvider } from 'react-intl';
import createHistory from 'history/createMemoryHistory';

import ProjectsPage from '../index';
import configureStore from '../../../configureStore';
import { translationMessages } from '../../../i18n';
import { mockLocalStorage, unmockLocalStorage } from '../../../support/tests/stub';

const history = createHistory();
const store = configureStore({}, history);

describe('<ProjectsPage />', () => {
  beforeEach(mockLocalStorage);
  afterEach(unmockLocalStorage);

  it('renders', () => {
    const renderedComponent = mount(
      <Provider store={store}>
        <IntlProvider locale="en" messages={translationMessages.en}>
          { React.createElement(ProjectsPage) }
        </IntlProvider>
      </Provider>
    );
    expect(renderedComponent.html()).toMatchSnapshot();
  });
});
