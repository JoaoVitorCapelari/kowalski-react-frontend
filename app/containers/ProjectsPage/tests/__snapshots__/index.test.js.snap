// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ProjectsPage /> renders 1`] = `"<div class=\\"kowalski-react-basic-container ProjectsPage__MainContainerWrapper-WbKmc jXcUyd\\"><!-- react-empty: 2 --><div class=\\"ProjectsPage__TitleBar-fqHHBt dGBGys\\"><h1 class=\\"ProjectsPage__PageTitle-NmeSH fIKzgA\\"><span>Projects</span></h1></div><div class=\\"ProjectsPage__ProjectListWrapper-kVbeTB jZprfF\\"><div class=\\"css-jxiqlq\\"><div class=\\"css-1lt1sep\\"></div><div class=\\"css-1vahj0q\\"></div></div></div><!-- react-empty: 9 --></div>"`;
